common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+    open: bool,
+}
+
+impl Bucket {
+    pub fn new(size: usize, rate: usize) -> Self {
+        assert!(rate > 0, "Fill rate must be greater than 0");
+        assert!(size > 0, "Bucket size must be greater than 0");
+        assert!(
+            size >= rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            tokens: size,
+            size,
+            rate,
+            last_refresh_time: Instant::now(),
+            open: false,
+        }
+    }
+
+    /// A fully open rate limiter, to allow for ignoring rate limiting for tests
+    pub fn open() -> Self {
+        Self {
+            tokens: std::usize::MAX,
+            size: std::usize::MAX,
+            rate: std::usize::MAX,
+            last_refresh_time: Instant::now(),
+            open: true,
+        }
+    }
+
+    /// Refill tokens based on how many seconds have passed since last refresh
+    pub(crate) fn refill(&mut self) {
+        let num_intervals = self.last_refresh_time.elapsed().as_secs();
+        if num_intervals > 0 {
+            self.add_tokens(num_intervals as usize * self.rate);
+            self.last_refresh_time = Instant::now();
+        }
+    }
+
+    /// Returns `usize` of tokens allowed.  May be less than requested.
+    /// For best effort, caller should return unused tokens with `add_tokens`
+    pub fn acquire_tokens(&mut self, count: usize) -> usize {
+        // Skip over if we purposely have an open throttle
+        if self.open {
+            return count;
+        }
+
+        // Refill if needed
+        self.refill();
+
+        let tokens = self.tokens;
+        let tokens_left = tokens.saturating_sub(count);
+        let tokens_allowed = min(tokens, count);
+        self.tokens = tokens_left;
+
+        tokens_allowed
+    }
+
+    /// Add new tokens, this can also be used to add unused tokens back
+    /// Ensures bucket doesn't overfill
+    pub fn add_tokens(&mut self, count: usize) {
+        self.tokens = min(self.size, self.tokens.saturating_add(count));
+    }
+}
+
+/// A rate limiter for `AsyncRead` interfaces to rate limit read bytes
+///
+/// This will pause and wait to send any future bytes until it's permitted to in the future
+pub struct AsyncReadRateLimiter<T: AsyncRead + Unpin> {
+    reader: T,
+    rate_limiter: Arc<Mutex<Bucket>>,
+}
+
+impl<T: AsyncRead + Unpin> AsyncReadRateLimiter<T> {
+    pub fn new(reader: T, rate_limiter: Arc<Mutex<Bucket>>) -> Self {
+        Self {
+            reader,
+            rate_limiter,
+        }
+    }
+}
+
+impl<T: AsyncRead + Unpin> AsyncRead for AsyncReadRateLimiter<T> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let requested = buf.len();
+        let allowed = Pin::new(&mut self.rate_limiter.lock().unwrap()).acquire_tokens(requested);
```

Comment:
the `Pin::new` stuff isn't necessary. should just be able to do `self.rate_limiter.lock().acquire_tokens(requested)`